### PR TITLE
Introduce initial autopilot binary

### PR DESCRIPTION
**Pull request by vkgnosis**
Has been discussed in https://github.com/cowprotocol/services/issues/175 and https://github.com/cowprotocol/services/issues/230 .

Currently does nothing except export a metric. Once it is integrated into kubernetes I will start moving functionality from Orderbook into it.

One design idea is that this can automatically be run by the orderbook (and the e2e tests) which is why it exports the main function in its library. I'm not yet sure what the best approach for that is with regards to its command line arguments. It seems like for anything that we want configurable (not default) we would need to duplicate them into the orderbook arguments. Or maybe there is some advanced Clap use that makes this easier (subcommands?). Anyway I don't want the details of this to delay other changes so will think about it more later.

Another thing to think about is that before the shared crate contained things needed for both orderbook and solver but now that we have 3 binaries using it this doesn't make much sense because we will also have things shared between all only two of the binaries. I think we should split up shared into more smaller crates which likely helps with compilation time too. Again I don't want this to delay the main change more so for now I will just move some stuff from orderbook into shared.

### Test Plan

CI and manually ran the autopilot binary to see that the metrics work

### CODE REVIEW

**Comment by nlordell** [comment]
> Or maybe there is some advanced Clap use that makes this easier (subcommands?). Anyway I don't want the details of this to delay other changes so will think about it more later.

One thing that I have thought about in the past would be to consider switching to configuration files instead of command line arguments. I would suggest maybe doing something like:

- `autopilot/src/lib.rs`
  ```rust
  #[derive(Default)]
  pub struct Configuration {
      // ...
  }
  
  pub struct Autopilot;
  
  impl Autopilot {
      pub async fn init(config: Configuration) -> Result<Self> {
          // ...
      }
      
      pub async fn start(self) -> Result<()> {
          // ...
      }
  }
  ```
- `autopilot/src/main.rs`
  ```rust
  #[derive(Default)]
  pub struct Configuration {
      // ...
  }
  
  pub struct Autopilot;
  #[tokio::main]
  pub async fn main() -> Result<()> {
      let config = read_from_file()?;
      let autopilot = Autopilot::init(config).await?;
      autopilot.start().await
  }
  ```
- `e2e/src/services.rs`
  ```rust
  pub async fn services() -> Result<Services> {
      let config = autopilot::Configuration {
          overrides_for_tests: true,
          ..Default::default()
      };
      let autopilot = Autopilot::init(config).await?;
      
      // Same for driver
      
      Ok(Services { autopilot, driver })
  }
  ```
  
The other advantage of configuration files is that it allows for richer configuration:
```toml
[[solver]]
kind = "http"
account = ...
http_specific_parameters = true

[[solver]]
kind = "baseline"
account = ...
baseline_specific_parameters = true

# ...
```

>  think we should split up shared into more smaller crates which likely helps with compilation time too.

I think this is a good idea.

**Review by nlordell — APPROVED** [review]
The start of something new!

File: crates/autopilot/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+struct Liveness;
+#[async_trait::async_trait]
+impl LivenessChecking for Liveness {
+    async fn is_alive(&self) -> bool {
+        true
+    }
+}
+
+/// Assumes tracing and metrics registry have already been set up.
+pub async fn main(args: arguments::Arguments) {
```
[the remark this answers is not in the document]
Comment (vkgnosis) [reply crates/autopilot/src/lib.rs:24]: I'll keep this in mind. Will do it as it comes naturally when I move more components into it.

**Comment by vkgnosis** [comment]
Feel free to leave more comments. I'm merging now so I get the docker image built with the binary and do the kubernetes setup.

**Pull request merged**